File: FlowCrypt/src/main/java/com/flowcrypt/email/api/retrofit/node/RequestsManager.java
package com.flowcrypt.email.api.retrofit.node;

import android.content.Context;
import android.net.Uri;
import android.os.AsyncTask;

import com.flowcrypt.email.api.retrofit.request.node.DecryptFileRequest;
import com.flowcrypt.email.api.retrofit.request.node.DecryptMsgRequest;
import com.flowcrypt.email.api.retrofit.request.node.EncryptFileRequest;
import com.flowcrypt.email.api.retrofit.request.node.EncryptMsgRequest;
import com.flowcrypt.email.api.retrofit.request.node.NodeRequest;
import com.flowcrypt.email.api.retrofit.request.node.NodeRequestWrapper;
import com.flowcrypt.email.api.retrofit.request.node.VersionRequest;
import com.flowcrypt.email.api.retrofit.response.node.BaseNodeResult;
import com.flowcrypt.email.api.retrofit.response.node.NodeResponseWrapper;
import com.flowcrypt.email.jetpack.livedata.SingleLiveEvent;
import com.flowcrypt.email.model.PgpKeyInfo;
import com.flowcrypt.email.node.NodeSecret;
import com.flowcrypt.email.node.TestData;

import java.util.Arrays;

import androidx.lifecycle.LiveData;
import retrofit2.Response;

/**
 * @author DenBond7
 */
public class RequestsManager {
  private SingleLiveEvent<NodeResponseWrapper> data;
  private NodeRetrofitHelper retrofitHelper;

  public RequestsManager(NodeSecret nodeSecret) {
    this.data = new SingleLiveEvent<>();
    this.retrofitHelper = NodeRetrofitHelper.getInstance(nodeSecret);
  }

  public LiveData<NodeResponseWrapper> getData() {
    return data;
  }

  public void getVersion(int requestCode) {
    load(requestCode, new VersionRequest());
  }

  public void encryptMsg(int requestCode, String msg) {
    load(requestCode, new EncryptMsgRequest(msg, Arrays.asList(TestData.getMixedPubKeys())));
  }

  public void decryptMsg(int requestCode, String msg, PgpKeyInfo[] prvKeys) {
    load(requestCode, new DecryptMsgRequest(msg, prvKeys, TestData.passphrases()));
  }

  public void encryptFile(int requestCode, byte[] data) {
    load(requestCode, new EncryptFileRequest(data, "file.txt", Arrays.asList(TestData.getMixedPubKeys())));
  }

  public void encryptFile(int requestCode, Context context, Uri fileUri) {
    load(requestCode, new EncryptFileRequest(context, fileUri, "file.txt", Arrays.asList(TestData.getMixedPubKeys())));
  }

  public void decryptFile(int requestCode, byte[] encryptedData, PgpKeyInfo[] prvKeys) {
    load(requestCode, new DecryptFileRequest(encryptedData, prvKeys, TestData.passphrases()));
  }

  private void load(final int requestCode, NodeRequest nodeRequest) {
    new Worker(data, retrofitHelper).execute(new NodeRequestWrapper<>(requestCode, nodeRequest));
  }

  private static class Worker extends AsyncTask<NodeRequestWrapper, Void, NodeResponseWrapper> {
    private SingleLiveEvent<NodeResponseWrapper> data;
    private NodeRetrofitHelper retrofitHelper;

    Worker(SingleLiveEvent<NodeResponseWrapper> data, NodeRetrofitHelper retrofitHelper) {
      this.data = data;
      this.retrofitHelper = retrofitHelper;
    }

    @Override
    protected NodeResponseWrapper doInBackground(NodeRequestWrapper... nodeRequestWrappers) {
      NodeRequestWrapper nodeRequestWrapper = nodeRequestWrappers[0];
      BaseNodeResult baseNodeResult;

      NodeService nodeService = retrofitHelper.getRetrofit().create(NodeService.class);

      try {
        Response response = nodeRequestWrapper.getRequest().getResponse(nodeService);
        if (response != null) {
          long time = response.raw().receivedResponseAtMillis() - response.raw().sentRequestAtMillis();
          if (response.body() != null) {
            baseNodeResult = (BaseNodeResult) response.body();
            baseNodeResult.setTime(time);
          } else {
            throw new NullPointerException("The response body is null!");
          }

          baseNodeResult.setTime(time);
        } else {
          throw new NullPointerException("The response is null!");
        }

      } catch (Exception e) {
        e.printStackTrace();
        return new NodeResponseWrapper<>(nodeRequestWrapper.getRequestCode(), e, (BaseNodeResult) null);
      }

      return new NodeResponseWrapper<>(nodeRequestWrapper.getRequestCode(), null, baseNodeResult);
    }

    @Override
    protected void onPostExecute(NodeResponseWrapper nodeResponseWrapper) {
      super.onPostExecute(nodeResponseWrapper);
      data.setValue(nodeResponseWrapper);
    }
  }
}

File: FlowCrypt/src/main/java/com/flowcrypt/email/api/retrofit/node/RequestsManager.kt
/*
 * © 2016-present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.api.retrofit.node

import android.content.Context
import android.net.Uri
import android.os.AsyncTask
import androidx.lifecycle.LiveData
import com.flowcrypt.email.api.retrofit.request.node.NodeRequest
import com.flowcrypt.email.api.retrofit.request.node.NodeRequestWrapper
import com.flowcrypt.email.api.retrofit.request.node.VersionRequest
import com.flowcrypt.email.api.retrofit.response.node.BaseNodeResponse
import com.flowcrypt.email.api.retrofit.response.node.NodeResponseWrapper
import com.flowcrypt.email.database.entity.KeyEntity
import com.flowcrypt.email.jetpack.livedata.SingleLiveEvent

/**
 * @author DenBond7
 */
//todo-denbond7 It's old code. We should drop this file
@Deprecated("We should drop this file")
object RequestsManager {
  private var data: SingleLiveEvent<NodeResponseWrapper<*>> = SingleLiveEvent()

  fun getData(): LiveData<NodeResponseWrapper<*>>? {
    return data
  }

  fun getVersion(requestCode: Int) {
    load(requestCode, VersionRequest())
  }

  fun encryptMsg(requestCode: Int, msg: String) {

  }

  fun decryptMsg(requestCode: Int, data: ByteArray = ByteArray(0), uri: Uri? = null,
                 prvKeys: Array<KeyEntity>, isEmail: Boolean = false) {

  }

  fun encryptFile(requestCode: Int, data: ByteArray) {

  }

  fun encryptFile(requestCode: Int, context: Context, fileUri: Uri) {

  }

  private fun load(requestCode: Int, nodeRequest: NodeRequest) {
    Worker(data).execute(NodeRequestWrapper(requestCode, nodeRequest))
  }

  private class Worker(
      private val data: SingleLiveEvent<NodeResponseWrapper<*>>) : AsyncTask<NodeRequestWrapper<*>, Void,
      NodeResponseWrapper<*>>() {

    override fun doInBackground(vararg nodeRequestWrappers: NodeRequestWrapper<*>): NodeResponseWrapper<*> {
      val nodeRequestWrapper = nodeRequestWrappers[0]
      val baseNodeResult: BaseNodeResponse

      val nodeService = NodeRetrofitHelper.getRetrofit()!!.create(NodeService::class.java)
      var time = 0L
      try {
        val response = nodeRequestWrapper.request.getResponse(nodeService)
        time = response.raw().receivedResponseAtMillis - response.raw().sentRequestAtMillis
        if (response.body() != null) {
          baseNodeResult = response.body() as BaseNodeResponse
        } else {
          throw NullPointerException("The response body is null!")
        }

      } catch (e: Exception) {
        e.printStackTrace()
        return NodeResponseWrapper.exception(nodeRequestWrapper.requestCode, e, null, time)
      }

      return NodeResponseWrapper.success(nodeRequestWrapper.requestCode, baseNodeResult, time)
    }

    override fun onPostExecute(nodeResponseWrapper: NodeResponseWrapper<*>) {
      super.onPostExecute(nodeResponseWrapper)
      data.value = nodeResponseWrapper
    }
  }
}
